framework/instance_gen.sage

```diff
@@ -26,16 +27,18 @@ def initialize_from_LWE_instance(dbdd_class, n, q, m, D_e,
     S = diagonal_matrix(m * [s_e] + n * [s_s] + [0])
     # draw matrix A and define the lattice
     A = matrix([[randint(0, q) for _ in range(n)] for _ in range(m)])
-    B = build_LWE_lattice(-A, q)
+    B = build_LWE_lattice(-A, q) # primal
+    D = build_LWE_lattice(A/q, 1/q) # dual
     # draw the secrets
     s = vec([draw_from_distribution(D_s) for _ in range(n)])
     e = vec([draw_from_distribution(D_e) for _ in range(m)])
     # compute the public value t and build a target
     b = (s * A.T + e) % q
     tar = concatenate([b, [0] * n])
     B = kannan_embedding(B, tar)
+    D = kannan_embedding(D, concatenate([-b/q, [0] * n])).T
     u = concatenate([e, s, [1]])
-    return A, b, dbdd_class(B, S, mu, u, verbosity=verbosity)
+    return A, b, dbdd_class(B, S, mu, u, verbosity=verbosity, D=D, Bvol=m*log(q))
```

Comment:
Could it happen that the volume prediction is not perfectly correct ? e.g. A=0 mod q, or more subtle degeneracy cases ?

Reply:
(maybe this is stupid... The actual lattice equation is in systematic form [A|Id] x = 0 mod q, so the volume does not depends on A).

Reply:
I don't think it could happen. Indeed, the lattice basis is built with the format 
```sage
q*I, 0, 0
A^T, I, 0
b,   0, 1
```
so the content of A and b don't impact the determinant of the basis (product of the diagonal coefficients), _ie_ its volume.

Reply:
:+1: